### ISSUE TITLE
External Filter Submit on KeyPress in 3.0 ?

### ISSUE THREAD

**Issue by orzdk** [angular-ui/ng-grid#3261]
Default behaviour for the filters is that they fire on each keypress, also when using external filters. I would like to avoid hitting the data API on each letter, so i tried the timeout buffer solution, waiting 500ms after last keypress, but I'm not too keen on that solution either. Is there ANY way, unsupported or hackish or not, that I can catch the keycode in gridApi.core.on.filterChanged and only fire on a return keycode 13? Thanks ! : )

**Comment by PaulL1** [comment]
The filterChanged is triggered by a watch on the term itself, not by keypresses on the widget.

I've merged today an ability to create custom directives for filters, and in the example where I do a modal popup, the term is only set once the whole filter is entered.  That same approach could work for your situation.  http://ui-grid.info/docs/#/tutorial/306_custom_filters

**Comment by orzdk** [comment]
Many thanks Paul, I will try that approach. I ended up wrapping the grid in a form submitting to a function that picked up a scopeobject that was populated with the filterChanged event. Then I discarded the header-textbox solution altogether because I forgot I needed a searchbox for a column that is grouped out of the data : ) Will be needing this for another grid though so cool with a new angle to try.

**Comment by nkoterba** [comment]
+1 to this feature.

I've used several other "Grids" (e.g. Kendo UI Grid) and by default they filter only after ENTER key press.

See demo: http://demos.telerik.com/kendo-ui/grid/filter-row

I was thinking adding either a `grid` or `columnDef` option would be pretty beneficial:

`filterOnKeypress` = boolean or perhaps an array of keycodes?

I understand that the filterChanged event is triggered by an Ng watch, but if an onkeydown event handler was added to the input of the Filter Header Template if `filterOnKeypress` = true, then another event could be fired, perhaps something like:
`filterKeyDown` or even `filterChangedByKeypress` with the actual value and/or key pressed.  This may also be more performant as its not using an Ng watch.

Just brainstorming ideas here.  I'm going to be implementing a custom directive as suggested above, but I wonder if the above enhancement wouldn't be too difficult and would help those displaying/filtering large amounts of data (who don't necessarily want the filter to be applied until the user has fully typed out their needed term).
